File: src/fensu/rules/layers/_helpers/dead_code.py
"""Invoke native production reachability policy from the Python project-rule host."""

import fensu._native as native
from fensu.analysis.models import PythonReachabilityFacts, PythonSymbolFact
from fensu.analysis.types import PythonSymbolKind
from fensu.rules.authoring.models import Fault, Project
from fensu.rules.authoring.types import RuleContext
from fensu.rules.layers.types import LayerCode


def check_unreachable_definition(*, project: Project, ctx: RuleContext) -> list[Fault]:
    """Evaluate definition reachability through inspectable, module-free project facts."""

    return _check_dead_code(project=project, ctx=ctx, code=LayerCode.UNREACHABLE_DEFINITION)


def check_stale_dead_code_root(*, project: Project, ctx: RuleContext) -> list[Fault]:
    """Evaluate configured roots through inspectable, module-free project facts."""

    return _check_dead_code(project=project, ctx=ctx, code=LayerCode.STALE_DEAD_CODE_ROOT)


def check_unreachable_module(*, project: Project, ctx: RuleContext) -> list[Fault]:
    """Evaluate module execution through inspectable, module-free project facts."""

    return _check_dead_code(project=project, ctx=ctx, code=LayerCode.UNREACHABLE_MODULE)


def _check_dead_code(*, project: Project, ctx: RuleContext, code: LayerCode) -> list[Fault]:
    """Keep project diagnostics executable even with an empty selected source surface."""

    del project
    if not ctx.project.dead_code.enabled:
        return []
    facts: PythonReachabilityFacts = ctx.project.python_reachability()
    modules: list[str] = sorted({symbol.module for symbol in facts.symbols})
    module_indexes: dict[str, int] = {module: index for index, module in enumerate(modules)}
    dead, stale = native.evaluate_python_reachability(
        (
            modules,
            [
                (module_indexes[symbol.module], symbol.name, symbol.kind is PythonSymbolKind.LOCAL)
                for symbol in facts.symbols
            ],
            [(edge.source.value, edge.target.value) for edge in facts.references],
            [entry.symbol.value for entry in facts.entrypoints if entry.symbol is not None],
        ),
        [(list(root.modules), list(root.symbols)) for root in ctx.project.dead_code.roots],
    )
    if code == LayerCode.STALE_DEAD_CODE_ROOT:
        return [
            ctx.fault_for(
                path=facts.configuration_path,
                line=1,
                column=0,
                message=(
                    f"dead_code.roots entry {index + 1} matches no existing production declaration"
                ),
                remediation=(
                    "Remove or correct the stale configured root; "
                    "matching does not depend on current reachability."
                ),
            )
            for index in stale
        ]
    modules_only: bool = code == LayerCode.UNREACHABLE_MODULE
    return [
        _declaration_fault(ctx=ctx, symbol=facts.symbols[index])
        for index in dead
        if (facts.symbols[index].kind is PythonSymbolKind.MODULE) == modules_only
    ]


def _declaration_fault(*, ctx: RuleContext, symbol: PythonSymbolFact) -> Fault:
    qualified: str = symbol.module if not symbol.name else f"{symbol.module}.{symbol.name}"
    return ctx.fault_at(
        location=symbol.location,
        message=f"unreachable {symbol.kind.value} {qualified}",
        remediation=(
            "Investigate the production entry mechanism before deleting code. "
            "Export an intentional public API or configure a reasoned root for dynamic dispatch."
        ),
    )

File: crates/fensu-native/src/rules/_helpers/dead_code/dead_code.rs
//! Project-owned reachability faults consumed by both check hosts.

use fensu_facts::dead_code::main::analyze::analyze;
use fensu_facts::dead_code::models::{Module, ProjectEntryPoint, Root};

use crate::rules::constants::{DEAD_DEFINITION_CODE, DEAD_MODULE_CODE, STALE_DEAD_CODE_ROOT_CODE};
use crate::rules::models::{NativeDeadCodeContext, NativeFaultRow, NativeProjectPlane};

const TEST_SCOPE: &str = "test";
const INIT_FILE: &str = "__init__.py";

pub(crate) fn faults(
    codes: &[String],
    context: &NativeDeadCodeContext,
    project: &NativeProjectPlane,
) -> Vec<NativeFaultRow> {
    if !context.enabled
        || !codes
            .iter()
            .any(|code| matches!(code.as_str(), "FFL106" | "FFL107" | "FFL108"))
    {
        return Vec::new();
    }
    let production = project
        .modules
        .iter()
        .filter(|module| module.scope != TEST_SCOPE && module.path.ends_with(".py"))
        .collect::<Vec<_>>();
    let modules = production
        .iter()
        .map(|module| {
            let initializer = module.path.ends_with("/__init__.py") || module.path == INIT_FILE;
            Module {
                name: module.module_parts.join("."),
                program: module.program.clone(),
                package: initializer && module.module_parts.len() == 1,
                initializer,
            }
        })
        .collect::<Vec<_>>();
    let roots = context
        .roots
        .iter()
        .map(|(modules, symbols)| Root {
            modules: modules.clone(),
            symbols: symbols.clone(),
        })
        .collect::<Vec<_>>();
    let entries = context
        .entrypoints
        .iter()
        .map(|(kind, reference)| ProjectEntryPoint {
            kind: kind.clone(),
            reference: reference.clone(),
        })
        .collect::<Vec<_>>();
    let analysis = analyze(&modules, &entries, &roots);
    let mut faults: Vec<NativeFaultRow> = Vec::new();
    if codes.iter().any(|code| code == STALE_DEAD_CODE_ROOT_CODE) {
        for index in analysis.stale_roots {
            faults.push(NativeFaultRow {
                code: STALE_DEAD_CODE_ROOT_CODE.to_owned(),
                line: 1,
                column: 0,
                message: Some(format!("dead_code.roots entry {} matches no existing production declaration", index + 1)),
                remediation: Some("Remove or correct the stale configured root; matching does not depend on current reachability.".to_owned()),
                path: Some(context.config_path.clone()),
            });
        }
    }
    for declaration in analysis.dead {
        let code = if declaration.name.is_empty() {
            DEAD_MODULE_CODE
        } else {
            DEAD_DEFINITION_CODE
        };
        if !codes.iter().any(|selected| selected == code) {
            continue;
        }
        let separator = if declaration.name.is_empty() { "" } else { "." };
        faults.push(NativeFaultRow {
            code: code.to_owned(),
            line: declaration.line,
            column: declaration.column,
            message: Some(format!("unreachable {} {}{separator}{}", declaration.kind, modules[declaration.module].name, declaration.name)),
            remediation: Some("Investigate the production entry mechanism before deleting code. Export an intentional public API or configure a reasoned root for dynamic dispatch.".to_owned()),
            path: Some(production[declaration.module].path.clone()),
        });
    }
    faults
}
